### ISSUE TITLE
Does pippo support @postconstruct and @predestroy annotations?

### ISSUE THREAD

**Issue by deepakmn** [pippo-java/pippo#508]
I'm looking at support for javax.annotations but couldn't find them used in any examples? Are these annotations, esp. PostConstruct and PreDestroy, supported?

**Comment by decebals** [comment]
Can you add more details about your use case? Do you talk in the context of `Spring Framework`? Do you talk in the context of [Controller](http://www.pippo.ro/mod/controller.html)?

**Comment by deepakmn** [comment]
Yes, in context of controller. I have an init method which is annotated with @PostConstruct but it does not seem to be invoked.

**Comment by decebals** [comment]
That annotations are used in the context of DI (dependency injection) and they are functional (out of the box) if you use [Spring](http://www.pippo.ro/mod/spring.html), [Guice](http://www.pippo.ro/mod/guice.html) or [CDI](http://www.pippo.ro/mod/cdi.html). 
On the other hand I don't think that is difficult to implement yourself a such functionality. All you must to do is to implement a custom [ControllerFactory](https://github.com/pippo-java/pippo/blob/master/pippo-controller-parent/pippo-controller/src/main/java/ro/pippo/controller/ControllerFactory.java) (functional interface - one method) that reads annotations from the supplied class via reflection, and after the instance is created, it calls (again via reflection) the method(s) annotated with `@PostConstruct`. Don't forget to add your custom controller factory using `Application.setControllerFactory`. An easy and enjoyable task, in my opinion.

**Comment by deepakmn** [comment]
Thanks for the insight. I didn't quite understand your question first, since you mentioned Spring framework. I'm using CDI in my app. Now that you mention that these annotations work out of the box, I wonder what I'm missing. It's a simple app with just 4 classes. I've placed beans.xml under src/main/resources/META-INF

The application:
```
public class JTapiApplication extends ControllerApplication {
	@Override
	protected void onInit() {
		Weld weld = new Weld();
        WeldContainer cdiContainer = weld.initialize();

        // registering WeldControllerFactory
        setControllerFactory(new WeldControllerFactory(cdiContainer));
        
		addControllers(new CallController());
	}
}
```
The controller
```
public class CallController extends Controller {
	@Inject
	private CallService callService; // Not injected

        @PostConstruct
	public void init() { // Not invoked
		System.out.println("init called");
	}

	@GET("call")
	public String placeCall(@Param("caller") String caller, @Param("callee") String callee) {
		if(callService != null) {  // callService is null, not injected
			callService.createCall(caller, callee);
		}
		return "This will place a jtapi call";
	}
	
	@GET("hangup")
	public String hangup(@Param("extn") String extn) {
		return "This will disconnect a jtapi call";
	}
}
```
The service
```
public class CallService {
	@Inject
	private AppConfig appConfig;
	
	@PostConstruct
	public void init() {
		//TODO: Init Jtapipeer and provider
		System.out.println("init called with " + appConfig);
	}
	
	public void createCall(String caller, String... callee) {
		System.out.println("Creating jtapi call");
	}
	
	public void disconnectCall(String extn) {
		System.out.println("Disconnecting jtapi call");
	}
}
```
Util class
```
public class AppConfig implements Serializable {
	private Properties appConfig;
	
	public AppConfig() throws IOException {
		appConfig = new Properties();
		load();
	}
	
	public Properties properties() {
		return appConfig;
	}
	
	public void reload() throws IOException {
		appConfig.clear();
		load();
	}
	
	private void load() throws IOException {
		appConfig.load(Files.newBufferedReader(Paths.get(".", "application.properties")));
	}
}
```

**Comment by decebals** [comment]
Share beans.xml. I don't understand where you define controllerService in cdi.

**Comment by deepakmn** [comment]
<beans xmlns="http://xmlns.jcp.org/xml/ns/javaee"
       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee 
		http://xmlns.jcp.org/xml/ns/javaee/beans_1_1.xsd"
       bean-discovery-mode="all">
</beans>

**Comment by deepakmn** [comment]
```
<beans xmlns="http://xmlns.jcp.org/xml/ns/javaee"
       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee 
		http://xmlns.jcp.org/xml/ns/javaee/beans_1_1.xsd"
       bean-discovery-mode="all">
</beans>
```

**Comment by decebals** [comment]
And where is declared `CallService` as bean?
Please see [pippo-demo-weld](https://github.com/pippo-java/pippo-demo/tree/master/pippo-demo-weld) and take a look at [WeldBeanProducer](https://github.com/pippo-java/pippo-demo/blob/master/pippo-demo-weld/src/main/java/ro/pippo/demo/weld/WeldBeanProducer.java) where `ContactService` is declared as bean.

**Comment by deepakmn** [comment]
Hey, I finally got it to work by adding the bean instance obtained from weldcontainer.

```
CallController cc = cdiContainer.select(CallController.class).get();
addControllers(cc);
```
When I created the producer class as you suggested, weld threw an ambiguous bean exception. I realized that weld was creating the bean for CallService without a producer class but was not available for the CallController unless we obtained the controller class as a bean. After this change, @PostConstruct is also working.
